File: shield_backend/scripts/test-welink.ts
/**
 * 测试 WeLink Webhook 发送消息
 * 用法: npx ts-node scripts/test-welink.ts
 */
const WEBHOOK_URL = "https://open.welink.huaweicloud.com/api/werobot/v1/webhook/send?token=***&channel=standard";

interface WeLinkResponse {
  code: string;
  data: string;
  message: string;
}

async function sendWelinkMessage(
  content: string,
  options?: {
    isAt?: boolean;
    isAtAll?: boolean;
    atAccounts?: string[];
  }
): Promise<WeLinkResponse | null> {
  // WeLink 要求消息长度 1~500
  if (content.length > 500) {
    content = content.slice(0, 497) + "...";
  }

  const body: Record<string, any> = {
    messageType: "text",
    content: {
      text: content,
    },
    timeStamp: Date.now(),
    uuid: crypto.randomUUID().replace(/-/g, ""),
    isAt: options?.isAt ?? false,
    isAtAll: options?.isAtAll ?? false,
  };

  if (options?.isAt && options?.atAccounts && options.atAccounts.length > 0) {
    body.atAccounts = options.atAccounts.slice(0, 10); // 最多10个
  }

  const res = await fetch(WEBHOOK_URL, {
    method: "POST",
    headers: {
      "Content-Type": "application/json; charset=UTF-8",
    },
    body: JSON.stringify(body),
  });

  const data: WeLinkResponse = await res.json();
  console.log("Status:", res.status);
  console.log("Response:", JSON.stringify(data, null, 2));
  return data;
}

// 测试发送
async function main() {
  const ip = await fetch("https://api.ipify.org", { timeout: 5000 })
    .then((r) => r.text())
    .catch(() => "unknown");

  const msg = `【EHS测试】这是一条来自安全联锁系统的测试消息 🧪\n时间：${new Date().toLocaleString("zh-CN")}\n发送IP：${ip}`;

  console.log("发送消息到 WeLink...");
  console.log("本机公网IP:", ip);
  console.log("消息长度:", msg.length, "字符\n");

  const result = await sendWelinkMessage(msg);

  if (result) {
    switch (result.code) {
      case "0":
        console.log("\n✅ 发送成功！去群里看看有没有消息。");
        break;
      case "58404":
        console.log("\n❌ 机器人资源不存在，检查 token 是否正确。");
        break;
      case "58601":
        console.log(`\n❌ 参数错误: ${result.message}`);
        break;
      case "58602":
        console.log("\n❌ 机器人未启用，检查 WeLink 群机器人状态。");
        break;
      default:
        console.log(`\n⚠️ 未知错误码 ${result.code}: ${result.message}`);
    }
  } else {
    console.log("\n❌ 发送失败，检查 IP 白名单和 token 是否正确。");
  }
}

main().catch(console.error);

File: shield_backend/scripts/test_welink.py
#!/usr/bin/env python3
"""
测试 WeLink Webhook 发送群消息
用法: python3 test_welink.py
"""

import urllib.request
import urllib.error
import json
import uuid
import time

# ⚠️ 把下面的 fc49bc…5f5d 换成你真实的完整 token
WEBHOOK_URL = "https://open.welink.huaweicloud.com/api/werobot/v1/webhook/send?token=fc49bc…5f5d&channel=standard"


def send_text(content: str, is_at: bool = False, is_at_all: bool = False, at_accounts: list = None):
    """
    发送文本消息到 WeLink 群
    
    Args:
        content: 消息内容（1~500字符）
        is_at: 是否@某个人
        is_at_all: 是否@全员
        at_accounts: 被@人员的userid列表，如 ["mettjhfuukaq@562a847505"]
    """
    # WeLink 要求消息长度 1~500
    if len(content) > 500:
        content = content[:497] + "..."
    
    body = {
        "messageType": "text",
        "content": {
            "text": content
        },
        "timeStamp": int(time.time() * 1000),
        "uuid": str(uuid.uuid4()).replace("-", ""),
        "isAt": is_at,
        "isAtAll": is_at_all,
    }
    
    if is_at and at_accounts:
        body["atAccounts"] = at_accounts[:10]  # 最多10个
    
    data = json.dumps(body, ensure_ascii=False).encode("utf-8")
    
    req = urllib.request.Request(
        WEBHOOK_URL,
        data=data,
        headers={
            "Content-Type": "application/json; charset=UTF-8"
        },
        method="POST"
    )
    
    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            result = json.loads(resp.read().decode("utf-8"))
            print(f"Status: {resp.status}")
            print(f"Response: {json.dumps(result, indent=2, ensure_ascii=False)}")
            return result
    except urllib.error.HTTPError as e:
        print(f"HTTP Error {e.code}: {e.reason}")
        try:
            print(f"Body: {e.read().decode('utf-8')}")
        except:
            pass
        return None
    except Exception as e:
        print(f"Error: {e}")
        return None


if __name__ == "__main__":
    # 获取本机公网IP
    try:
        ip = urllib.request.urlopen("https://api.ipify.org", timeout=5).read().decode()
    except:
        ip = "unknown"
    
    msg = f"【EHS测试】这是一条来自安全联锁系统的测试消息 🧪\n时间：2026-05-28\n发送IP：{ip}"
    
    print(f"发送消息到 WeLink...")
    print(f"本机公网IP: {ip}")
    print(f"消息长度: {len(msg)} 字符\n")
    
    result = send_text(msg)
    
    if result:
        code = result.get("code")
        if code == "0":
            print("\n✅ 发送成功！去群里看看有没有消息。")
        elif code == "58404":
            print("\n❌ 机器人资源不存在，检查 token 是否正确。")
        elif code == "58601":
            print(f"\n❌ 参数错误: {result.get('message')}")
        elif code == "58602":
            print("\n❌ 机器人未启用，检查 WeLink 群机器人状态。")
        else:
            print(f"\n⚠️ 未知错误码 {code}: {result.get('message')}")
    else:
        print("\n❌ 发送失败，检查 IP 白名单和 token 是否正确。")
